python/ray/experimental/streaming/communication.py

```diff
@@ -0,0 +1,271 @@
+                slot = shuffle_by_key_destinations.setdefault(
+                    channel.dst_operator_id, index_2)
+                self.shuffle_key_channels[slot].append(channel)
+                if slot == index_2:
+                    index_2 += 1
+            else:  # TODO (john): Add support for other partitioting strategies
+                sys.exit("Unrecognized or unsupported partitioning strategy.")
+        # A KeyedDataStream can only be shuffled by key
+        assert (not (self.shuffle_exists and self.shuffle_key_exists))
+
+    # Flushes any remaining records in the output channels
+    # 'close' indicates whether we should also 'close' the channel (True) by propagating 'None'
+    # or just flush the remaining records to plasma (False)
+    def _flush(self, close=False):
+        """Flushes remaining output records in the output queues to plasma.
+
+         None is used as special type of record that is propagated from sources
+         to sink to notify that the end of data in a stream.
+
+         Attributes:
+             close (bool): A flag denoting whether the channel should be also marked
+             as 'closed' (True) or not (False) after flushing.
+        """
+        for channel in self.forward_channels:
+            if close == True:
+                channel.queue.put_next(None)
+            channel.queue._flush_writes()
+        for channels in self.shuffle_channels:
+            for channel in channels:
+                if close == True:
+                    channel.queue.put_next(None)
+                channel.queue._flush_writes()
+        for channels in self.shuffle_key_channels:
+            for channel in channels:
+                if close == True:
+                    channel.queue.put_next(None)
+                channel.queue._flush_writes()
+        # TODO (john): Add more channel types
+
+    # Returns all destination actor ids
+    def _destination_actor_ids(self):
+        destinations = []
+        for channel in self.forward_channels:
+            destinations.append((channel.dst_operator_id,
+                                 channel.dst_instance_id))
+        for channels in self.shuffle_channels:
+            for channel in channels:
+                destinations.append((channel.dst_operator_id,
+                                     channel.dst_instance_id))
+        for channels in self.shuffle_key_channels:
+            for channel in channels:
+                destinations.append((channel.dst_operator_id,
+                                     channel.dst_instance_id))
+        # TODO (john): Add more channel types
+        return destinations
+
+    # Pushes the record to the output
+    # Each individual output queue flushes batches to plasma periodically based on 'batch_max_size' and 'batch_max_time'
+    def _push(self, record):
+        # Forward record
+        for channel in self.forward_channels:
+            # print("[writer] Push record '{}' to channel {}".format(record,c))
```

Comment:
stuff like that should be done with the python logging module (then debug logging can conveniently be switched on and off)

Reply:
I've added a couple of tests (not in .travis.yml though)

Reply:
Sorry, I replied to the wrong comment. I now use a logger for all messages.